src/calliope/config.py

```diff
@@ -0,0 +1,369 @@
+    """
+
+    def _hide_from_schema(schema: dict):
+        for hide in to_hide:
+            schema.get("properties", {}).pop(hide, None)
+        return schema
+
+    return _hide_from_schema
+
+
+class ConfigBaseModel(BaseModel):
+    """A base class for creating pydantic models for Calliope configuration options."""
+
+    _kwargs: dict = {}
+
+    def update(self, update_dict: dict, deep: bool = False) -> Self:
+        """Return a new iteration of the model with updated fields.
+
+        Updates are validated and stored in the parent class in the `_kwargs` key.
+
+        Args:
+            update_dict (dict): Dictionary with which to update the base model.
+            deep (bool, optional): Set to True to make a deep copy of the model. Defaults to False.
+
+        Returns:
+            BaseModel: New model instance.
+        """
+        new_dict: dict = {}
+        # Iterate through dict to be updated and convert any sub-dicts into their respective pydantic model objects
+        for key, val in update_dict.items():
+            key_class = getattr(self, key)
+            if isinstance(key_class, ConfigBaseModel):
+                new_dict[key] = key_class.update(val)
+                key_class._kwargs = val
+            else:
+                new_dict[key] = val
+        updated = super().model_copy(update=new_dict, deep=deep)
+        updated.model_validate(updated)
+        self._kwargs = update_dict
+        return updated
+
+    @overload
+    def model_yaml_schema(self, filepath: str | Path) -> None: ...
+
+    @overload
+    def model_yaml_schema(self, filepath: None = None) -> str: ...
+
+    def model_yaml_schema(self, filepath: str | Path | None = None) -> None | str:
+        """Generate a YAML schema for the class.
+
+        Args:
+            filepath (str | Path | None, optional): If given, save schema to given path. Defaults to None.
+
+        Returns:
+            None | str: If `filepath` is given, returns None. Otherwise, returns the YAML string.
+        """
+        # By default, the schema uses $ref/$def cross-referencing for each pydantic model class,
+        # but this isn't very readable when rendered in our documentation.
+        # So, we resolve references and then delete all the `$defs`
+        schema_dict = AttrDict(jsonref.replace_refs(self.model_json_schema()))
+        schema_dict.del_key("$defs")
+        return schema_dict.to_yaml(filepath)
+
+    @property
+    def applied_keyword_overrides(self) -> dict:
+        """Most recently applied keyword overrides used to update this configuration.
+
+        Returns:
+            dict: Description of applied overrides.
+        """
+        return self._kwargs
+
+
+class ModeBaseModel(ConfigBaseModel):
+    """Mode-specific configuration, which will be hidden from the string representation of the model if that mode is not activated."""
+
+    mode: MODES_T = Field(default="plan")
+    """Mode in which to run the optimisation."""
```

Comment:
These _are_ docstrings, see [this pydantic PR](https://github.com/pydantic/pydantic/pull/6563). You'll notice that if you write `model.config.build.mode` in VSCode and hover over it, it will give you the docstring in the tooltip thanks to intellisense.

Reply:
Huh. That's quite fancy. Alright, this is no issue then!